File: scripts/computer-use-fixtures/focus_probe.m
// focus-probe: samples the user's focus state at a fixed rate so a fixture or
// certification run can prove zero focus theft. Read-only: it never posts an
// event, activates an app, or touches another process's state. Output is
// newline-delimited JSON on stdout; analysis lives in
// apps/desktop/src/cuaFixtures/focusProbe.ts.
//
// Usage:
//   focus-probe --duration <seconds> [--hz <rate>] [--label <text>]
//   focus-probe --stdin [--duration <seconds>] [--hz <rate>]
//   focus-probe --once
//
// Options: --hz (default 50, 1-500), --top-win-every <n> (CGWindowList read
// every n-th tick; default 5), --label <text> recorded in the meta line.
//
// --stdin stops sampling when stdin reaches EOF, so a parent holds the pipe
// open for the measured section and closes it to stop. --duration is a hard
// cap in both modes. SIGINT/SIGTERM/SIGHUP stop cleanly.
//
// Exit codes: 0 clean stop, 1 usage error, 2 runtime failure.
//
// Sample line shape (keys absent fields are emitted as null):
//   {"t":<ms since start>,"pid":<frontmost pid>,"app":<frontmost name>,
//    "keyWin":<AX focused window CGWindowID>,"keyTitle":<its title>,
//    "topWin":<topmost layer-0 CGWindowID for front pid>,"topTitle":<title>,
//    "space":<SLSGetActiveSpace id>,"focused":"<pid>:<role>:<title>"}
//
// keyWin/focused need Accessibility trust (AXIsProcessTrusted on this binary);
// topWin/topTitle need nothing, and topTitle additionally needs a screen
// recording grant to carry the real title. space needs the private
// SLSGetActiveSpace symbol. Every field degrades to null independently.
#import <AppKit/AppKit.h>
#import <ApplicationServices/ApplicationServices.h>
#import <CoreGraphics/CoreGraphics.h>
#include <dlfcn.h>
#include <errno.h>
#include <limits.h>
#include <mach/mach_time.h>
#include <poll.h>
#include <signal.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

typedef void (*SLPSGetFrontProcessFn)(ProcessSerialNumber *psn);
typedef int (*CGSMainConnectionIDFn)(void);
typedef uint64_t (*SLSGetActiveSpaceFn)(int);
// Private HIServices bridge: AXUIElement -> CGWindowID. Resolved dynamically,
// same story as the SkyLight symbols in space_ctl.m.
typedef AXError (*AXUIElementGetWindowFn)(AXUIElementRef element, CGWindowID *windowId);

static volatile sig_atomic_t gSignalStop = 0;
static void onSignal(int signal) { (void)signal; gSignalStop = 1; }

static SLPSGetFrontProcessFn gGetFront;
static SLSGetActiveSpaceFn gGetActiveSpace;
static AXUIElementGetWindowFn gAXWindowId;
static int gConnection = -1;
static mach_timebase_info_data_t gTimebase;

static double monotonicMs(void) {
  return (double)(mach_absolute_time() * gTimebase.numer / gTimebase.denom) / 1e6;
}

static void jsonString(FILE *stream, const char *value) {
  fputc('"', stream);
  for (const unsigned char *cursor = (const unsigned char *)value; *cursor; cursor++) {
    switch (*cursor) {
      case '"': fputs("\\\"", stream); break;
      case '\\': fputs("\\\\", stream); break;
      case '\n': fputs("\\n", stream); break;
      case '\r': fputs("\\r", stream); break;
      case '\t': fputs("\\t", stream); break;
      default:
        if (*cursor < 0x20)
          fprintf(stream, "\\u%04x", *cursor);
        else
          fputc(*cursor, stream);
    }
  }
  fputc('"', stream);
}

static void jsonNumberOrNull(FILE *stream, bool present, long long value) {
  if (present)
    fprintf(stream, "%lld", value);
  else
    fputs("null", stream);
}

static void jsonTextOrNull(FILE *stream, const char *value) {
  if (value)
    jsonString(stream, value);
  else
    fputs("null", stream);
}

static void cfText(CFStringRef string, char *buffer, size_t size) {
  buffer[0] = 0;
  if (string)
    CFStringGetCString(string, buffer, size, kCFStringEncodingUTF8);
}

static pid_t frontmostPid(void) {
  // NSWorkspace's cached frontmost application needs a serviced run loop.
  // This sampler polls synchronously, so prefer the native current reading.
  if (gGetFront) {
    ProcessSerialNumber psn = {0, 0};
    gGetFront(&psn);
    pid_t pid = -1;
    if (GetProcessPID(&psn, &pid) == noErr && pid > 0)
      return pid;
  }
  // Only the fallback needs AppKit's cached observation refreshed.
  CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0, true);
  NSRunningApplication *front = NSWorkspace.sharedWorkspace.frontmostApplication;
  if (front)
    return front.processIdentifier;
  return -1;
}

static void frontmostName(pid_t pid, char *buffer, size_t size) {
  buffer[0] = 0;
  if (pid < 1)
    return;
  NSRunningApplication *app = [NSRunningApplication runningApplicationWithProcessIdentifier:pid];
  NSString *name = app.localizedName ?: app.bundleIdentifier;
  if (name)
    strncpy(buffer, name.UTF8String, size - 1);
}

// One AX round trip for a list of attributes; failed entries come back as
// AXValue-wrapped AXError and are skipped by the string/number readers.
static CFArrayRef copyAttributes(AXUIElementRef element, CFArrayRef attributes) {
  CFArrayRef values = NULL;
  if (AXUIElementCopyMultipleAttributeValues(element, attributes,
                                             (AXCopyMultipleAttributeOptions)0,
                                             &values) != kAXErrorSuccess ||
      !values || CFGetTypeID(values) != CFArrayGetTypeID())
    return NULL;
  return values;
}

static bool attrIsError(CFTypeRef value) {
  return !value || CFGetTypeID(value) != CFStringGetTypeID();
}

// AX focused ("key") window of the frontmost app + the system focused element.
// Both need this process to hold the Accessibility grant. focusedPid is the
// owner of the system-wide focused element: typing focus can move to another
// app (desktop, menu bar, agent target) while frontmostApplication still
// reads the human's app, so it is a first-class theft field.
static void axSample(pid_t frontPid, CGWindowID *keyWindow, char *keyTitle, size_t keyTitleSize,
                     pid_t *focusedPid, char *focused, size_t focusedSize) {
  *keyWindow = kCGNullWindowID;
  *focusedPid = -1;
  keyTitle[0] = 0;
  focused[0] = 0;
  if (!AXIsProcessTrusted())
    return;

  if (frontPid > 0) {
    AXUIElementRef app = AXUIElementCreateApplication(frontPid);
    if (app) {
      CFTypeRef window = NULL;
      if (AXUIElementCopyAttributeValue(app, kAXFocusedWindowAttribute, &window) == kAXErrorSuccess &&
          window && CFGetTypeID(window) == AXUIElementGetTypeID()) {
        if (gAXWindowId) {
          CGWindowID windowId = kCGNullWindowID;
          if (gAXWindowId((AXUIElementRef)window, &windowId) == kAXErrorSuccess)
            *keyWindow = windowId;
        }
        const void *names[] = {kAXTitleAttribute};
        CFArrayRef attributes =
            CFArrayCreate(NULL, names, 1, &kCFTypeArrayCallBacks);
        CFArrayRef values = copyAttributes((AXUIElementRef)window, attributes);
        if (values && CFArrayGetCount(values) == 1) {
          CFTypeRef title = CFArrayGetValueAtIndex(values, 0);
          if (!attrIsError(title))
            cfText((CFStringRef)title, keyTitle, keyTitleSize);
        }
        if (values)
          CFRelease(values);
        CFRelease(attributes);
        CFRelease(window);
      }
      CFRelease(app);
    }
  }

  AXUIElementRef system = AXUIElementCreateSystemWide();
  if (!system)
    return;
  CFTypeRef element = NULL;
  if (AXUIElementCopyAttributeValue(system, kAXFocusedUIElementAttribute, &element) ==
          kAXErrorSuccess &&
      element && CFGetTypeID(element) == AXUIElementGetTypeID()) {
    const void *names[] = {kAXRoleAttribute, kAXTitleAttribute, CFSTR("AXDescription"),
                           kAXIdentifierAttribute, CFSTR("AXPid")};
    CFArrayRef attributes = CFArrayCreate(NULL, names, 5, &kCFTypeArrayCallBacks);
    CFArrayRef values = copyAttributes((AXUIElementRef)element, attributes);
    char role[128] = {0};
    char label[192] = {0};
    if (values && CFArrayGetCount(values) == 5) {
      CFTypeRef roleValue = CFArrayGetValueAtIndex(values, 0);
      if (!attrIsError(roleValue))
        cfText((CFStringRef)roleValue, role, sizeof(role));
      for (CFIndex index = 1; index <= 3; index++) {
        CFTypeRef text = CFArrayGetValueAtIndex(values, index);
        if (!attrIsError(text)) {
          cfText((CFStringRef)text, label, sizeof(label));
          if (label[0])
            break;
        }
      }
      CFTypeRef pidValue = CFArrayGetValueAtIndex(values, 4);
      if (pidValue && CFGetTypeID(pidValue) == CFNumberGetTypeID())
        CFNumberGetValue((CFNumberRef)pidValue, kCFNumberIntType, focusedPid);
    }
    if (values)
      CFRelease(values);
    CFRelease(attributes);
    if (*focusedPid < 1)
      AXUIElementGetPid((AXUIElementRef)element, focusedPid);
    // A compact identity string: the analyzer only needs change detection.
    snprintf(focused, focusedSize, "%d:%s:%s", (int)*focusedPid, role, label);
    CFRelease(element);
  } else if (element) {
    CFRelease(element);
  }
  CFRelease(system);
}

// Topmost normal-layer onscreen window owned by the frontmost process. Needs
// no grant for the id; the title follows the screen-recording grant.
static void topWindowSample(pid_t frontPid, CGWindowID *window, char *title, size_t titleSize) {
  *window = kCGNullWindowID;
  title[0] = 0;
  if (frontPid < 1)
    return;
  CFArrayRef list = CGWindowListCopyWindowInfo(
      kCGWindowListOptionOnScreenOnly | kCGWindowListExcludeDesktopElements, kCGNullWindowID);
  if (!list)
    return;
  CFIndex count = CFArrayGetCount(list);
  for (CFIndex index = 0; index < count; index++) {
    CFDictionaryRef info = (CFDictionaryRef)CFArrayGetValueAtIndex(list, index);
    CFNumberRef owner = (CFNumberRef)CFDictionaryGetValue(info, kCGWindowOwnerPID);
    int ownerPid = 0;
    if (!owner || !CFNumberGetValue(owner, kCFNumberIntType, &ownerPid) || ownerPid != frontPid)
      continue;
    CFNumberRef layer = (CFNumberRef)CFDictionaryGetValue(info, kCGWindowLayer);
    int layerValue = -1;
    if (layer)
      CFNumberGetValue(layer, kCFNumberIntType, &layerValue);
    if (layerValue != 0)
      continue;
    CFNumberRef number = (CFNumberRef)CFDictionaryGetValue(info, kCGWindowNumber);
    int numberValue = 0;
    if (number && CFNumberGetValue(number, kCFNumberSInt32Type, &numberValue))
      *window = (CGWindowID)numberValue;
    cfText((CFStringRef)CFDictionaryGetValue(info, kCGWindowName), title, titleSize);
    break;
  }
  CFRelease(list);
}

static void emitSample(double tMs, bool withTopWindow) {
  pid_t pid = frontmostPid();
  char app[256];
  frontmostName(pid, app, sizeof(app));

  CGWindowID keyWindow = kCGNullWindowID;
  char keyTitle[512];
  pid_t focusedPid = -1;
  char focused[384];
  axSample(pid, &keyWindow, keyTitle, sizeof(keyTitle), &focusedPid, focused, sizeof(focused));

  // CGWindowListCopyWindowInfo costs ~5-10 ms per call, so this field runs on
  // a divisor of the tick rate. An unsampled tick emits null rather than a
  // stale repeat, so the analyzer never mistakes a repeat for a reading.
  CGWindowID topWindow = kCGNullWindowID;
  char topTitle[512];
  topTitle[0] = 0;
  if (withTopWindow)
    topWindowSample(pid, &topWindow, topTitle, sizeof(topTitle));

  bool spaceSet = false;
  uint64_t space = 0;
  if (gGetActiveSpace && gConnection >= 0) {
    space = gGetActiveSpace(gConnection);
    spaceSet = space != 0;
  }

  fprintf(stdout, "{\"t\":%.1f,\"pid\":", tMs);
  jsonNumberOrNull(stdout, pid > 0, pid);
  fputs(",\"app\":", stdout);
  jsonTextOrNull(stdout, app[0] ? app : NULL);
  fputs(",\"keyWin\":", stdout);
  jsonNumberOrNull(stdout, keyWindow != kCGNullWindowID, keyWindow);
  fputs(",\"keyTitle\":", stdout);
  jsonTextOrNull(stdout, keyTitle[0] ? keyTitle : NULL);
  fputs(",\"topWin\":", stdout);
  jsonNumberOrNull(stdout, topWindow != kCGNullWindowID, topWindow);
  fputs(",\"topTitle\":", stdout);
  jsonTextOrNull(stdout, topTitle[0] ? topTitle : NULL);
  fputs(",\"space\":", stdout);
  jsonNumberOrNull(stdout, spaceSet, (long long)space);
  fputs(",\"focusedPid\":", stdout);
  jsonNumberOrNull(stdout, focusedPid > 0, focusedPid);
  fputs(",\"focused\":", stdout);
  jsonTextOrNull(stdout, focused[0] ? focused : NULL);
  fputs("}\n", stdout);
}

static bool stdinClosed(void) {
  struct pollfd descriptor = {.fd = STDIN_FILENO, .events = POLLIN | POLLHUP};
  if (poll(&descriptor, 1, 0) <= 0)
    return false;
  if (!(descriptor.revents & (POLLIN | POLLHUP | POLLERR)))
    return false;
  char buffer[256];
  ssize_t count = read(STDIN_FILENO, buffer, sizeof(buffer));
  if (count == 0)
    return true;
  if (count < 0 && errno != EAGAIN && errno != EWOULDBLOCK && errno != EINTR)
    return true;
  return false;
}

// nanosleep/usleep/mach_wait_until all overshoot badly under macOS timer
// coalescing on this class of process (measured: 40-80% over the ask). poll()
// wakes within ~5% and a short spin closes the rest, so the tick cadence the
// report advertises is the cadence samples actually ran at.
static void waitUntil(double deadlineMs, double startEpochMs, uint64_t startTicks) {
  double remainingMs = deadlineMs - monotonicMs();
  if (remainingMs <= 0)
    return;
  // deadlineMs is on the monotonicMs() scale whose zero is startEpochMs;
  // convert the *delta* to mach ticks or the deadline lands eons out.
  uint64_t deadline = startTicks + (uint64_t)((deadlineMs - startEpochMs) * 1e6 *
                                            (double)gTimebase.denom / (double)gTimebase.numer);
  if (remainingMs > 3.0)
    poll(NULL, 0, (int)(remainingMs - 1.5));
  while (mach_absolute_time() < deadline) {
  }
}

static bool parseDouble(const char *text, double minimum, double maximum, double *value) {
  char *end = NULL;
  errno = 0;
  double parsed = strtod(text, &end);
  if (errno == ERANGE || end == text || *end != '\0' || parsed < minimum || parsed > maximum)
    return false;
  *value = parsed;
  return true;
}

int main(int argc, char **argv) {
  @autoreleasepool {
    double hz = 50.0;
    double durationSeconds = 0;
    bool durationSet = false;
    bool watchStdin = false;
    bool once = false;
    long topWinEvery = 5;
    const char *label = NULL;

    for (int index = 1; index < argc; index++) {
      const char *argument = argv[index];
      if (!strcmp(argument, "--stdin")) {
        watchStdin = true;
        continue;
      }
      if (!strcmp(argument, "--once")) {
        once = true;
        continue;
      }
      if (!strcmp(argument, "--hz") || !strcmp(argument, "--duration") ||
          !strcmp(argument, "--label") || !strcmp(argument, "--top-win-every")) {
        if (index + 1 >= argc) {
          fprintf(stderr, "focus-probe: %s needs a value\n", argument);
          return 1;
        }
        const char *value = argv[++index];
        if (!strcmp(argument, "--hz")) {
          if (!parseDouble(value, 1, 500, &hz)) {
            fprintf(stderr, "focus-probe: invalid --hz: %s\n", value);
            return 1;
          }
        } else if (!strcmp(argument, "--duration")) {
          if (!parseDouble(value, 0.05, 86400, &durationSeconds)) {
            fprintf(stderr, "focus-probe: invalid --duration: %s\n", value);
            return 1;
          }
          durationSet = true;
        } else if (!strcmp(argument, "--top-win-every")) {
          double parsed;
          if (!parseDouble(value, 1, 10000, &parsed)) {
            fprintf(stderr, "focus-probe: invalid --top-win-every: %s\n", value);
            return 1;
          }
          topWinEvery = (long)parsed;
        } else {
          label = value;
        }
        continue;
      }
      fprintf(stderr, "focus-probe: unknown argument: %s\n", argument);
      return 1;
    }
    if (!once && !watchStdin && !durationSet) {
      fprintf(stderr, "focus-probe: pass --duration, --stdin, or --once\n");
      return 1;
    }

    mach_timebase_info(&gTimebase);
    signal(SIGINT, onSignal);
    signal(SIGTERM, onSignal);
    signal(SIGHUP, onSignal);

    void *appServices = dlopen(
        "/System/Library/Frameworks/ApplicationServices.framework/ApplicationServices",
        RTLD_NOW | RTLD_GLOBAL);
    gGetFront = (SLPSGetFrontProcessFn)dlsym(RTLD_DEFAULT, "_SLPSGetFrontProcess");
    gAXWindowId = (AXUIElementGetWindowFn)dlsym(RTLD_DEFAULT, "_AXUIElementGetWindow");
    if (!gAXWindowId && appServices)
      gAXWindowId = (AXUIElementGetWindowFn)dlsym(appServices, "_AXUIElementGetWindow");

    void *sky = dlopen("/System/Library/PrivateFrameworks/SkyLight.framework/SkyLight",
                       RTLD_NOW | RTLD_GLOBAL);
    CGSMainConnectionIDFn connectionId = NULL;
    if (sky) {
      connectionId = (CGSMainConnectionIDFn)dlsym(sky, "CGSMainConnectionID");
      gGetActiveSpace = (SLSGetActiveSpaceFn)dlsym(sky, "SLSGetActiveSpace");
    }
    if (connectionId)
      gConnection = connectionId();

    setbuf(stdout, NULL);
    bool axTrusted = AXIsProcessTrusted();
    fprintf(stdout, "{\"kind\":\"meta\",\"pid\":%d,\"hz\":%.2f,\"topWinEvery\":%ld,"
                    "\"axTrusted\":%s,\"axWindowSymbol\":%s,\"slsSpace\":%s,"
                    "\"stdinWatch\":%s,\"label\":",
            getpid(), hz, topWinEvery, axTrusted ? "true" : "false",
            gAXWindowId ? "true" : "false",
            (gGetActiveSpace && gConnection >= 0) ? "true" : "false",
            watchStdin ? "true" : "false");
    jsonTextOrNull(stdout, label);
    fprintf(stdout, ",\"startedAt\":%.3f}\n",
            [[NSDate date] timeIntervalSince1970]);

    const double periodMs = 1000.0 / hz;
    const double durationMs = durationSeconds * 1000.0;
    const double started = monotonicMs();
    const uint64_t startTicks = mach_absolute_time();
    unsigned long long samples = 0;
    unsigned long long overruns = 0;
    double sampleMsTotal = 0;
    double sampleMsMax = 0;
    const char *stoppedBy = "duration";

    for (uint64_t tick = 0;; tick++) {
      const double now = monotonicMs();
      const double elapsed = now - started;
      if (durationSet && elapsed >= durationMs)
        break;
      if (gSignalStop) {
        stoppedBy = "signal";
        break;
      }
      if (watchStdin && stdinClosed()) {
        stoppedBy = "stdin";
        break;
      }
      @autoreleasepool {
        emitSample(elapsed, tick % (uint64_t)topWinEvery == 0);
      }
      samples++;
      const double emitMs = monotonicMs() - now;
      sampleMsTotal += emitMs;
      if (emitMs > sampleMsMax)
        sampleMsMax = emitMs;
      if (once) {
        stoppedBy = "once";
        break;
      }
      const double next = started + (double)(tick + 1) * periodMs;
      if (next - monotonicMs() < -periodMs)
        overruns++;
      else
        waitUntil(next, started, startTicks);
    }

    fprintf(stdout,
            "{\"kind\":\"done\",\"samples\":%llu,\"elapsedMs\":%.1f,\"overruns\":%llu,"
            "\"sampleMsAvg\":%.2f,\"sampleMsMax\":%.2f,\"stoppedBy\":",
            samples, monotonicMs() - started, overruns,
            samples ? sampleMsTotal / (double)samples : 0.0, sampleMsMax);
    jsonString(stdout, stoppedBy);
    fputs("}\n", stdout);
    return 0;
  }
}

File: scripts/computer-use-fixtures/focus-probe.ts
/**
 * focus-probe: standalone focus-theft meter for fixture and certification
 * runs. Wraps the focus_probe.m sampler binary (NDJSON on stdout) with the
 * analyzer from apps/desktop/src/cuaFixtures/focusProbe.ts and turns the
 * result into an assertable exit code plus a JSON report.
 *
 * Usage:
 *   bun scripts/computer-use-fixtures/focus-probe.ts --duration 30
 *   bun scripts/computer-use-fixtures/focus-probe.ts --wrap -- ./my-cert-script.sh
 *   bun scripts/computer-use-fixtures/focus-probe.ts --duration 30 \
 *     --expect-frontmost-pid 4242 --expect-key-window 9001 \
 *     --report report.json --samples-out samples.ndjson
 *
 * Exit codes: 0 theft-free, 1 usage/environment error, 2 off-baseline
 * samples detected (or coverage too thin to make the claim).
 *
 * The probe is read-only: NSWorkspace/SLPS for the frontmost pid, AX for the
 * key window and focused element, CGWindowList for the top window, and a
 * private SLSGetActiveSpace read for the Space. A null field means the
 * matching grant/symbol was unavailable; it narrows coverage, it never
 * counts as theft. Grant the *runner's* terminal/app Accessibility (and
 * Screen Recording for window titles) to measure keyWin and focused.
 */
import { spawn, spawnSync, type ChildProcessWithoutNullStreams } from "node:child_process";
import { existsSync } from "node:fs";
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  analyzeFocusSamples,
  parseFocusProbeLine,
  type FocusProbeDone,
  type FocusProbeExpect,
  type FocusProbeMeta,
  type FocusProbeSample,
} from "../../apps/desktop/src/cuaFixtures/focusProbe";

const root = fileURLToPath(new URL("../../", import.meta.url));
const defaultBinary = "/private/tmp/synara-cua-implementation/focus-probe";

interface CliOptions {
  binary: string;
  durationSeconds: number | null;
  wrap: string[] | null;
  hz: number;
  settleMs: number;
  expect: FocusProbeExpect;
  strictFocus: boolean;
  minSamples: number;
  reportPath: string | null;
  samplesPath: string | null;
  label: string;
}

const usage = `focus-probe.ts --duration <seconds> | --wrap -- <command...>
  --binary <path>            sampler binary (default ${defaultBinary}; built with clang when missing)
  --duration <seconds>       sample for a fixed duration
  --wrap -- <command...>     sample until the wrapped command exits
  --hz <rate>                samples per second (default 50, binary allows 1-500)
  --settle-ms <ms>           warm-up window that seeds the baseline (default 0)
  --expect-frontmost-pid <n> pin baseline frontmost pid (e.g. the human's app)
  --expect-key-window <n>    pin baseline AX key-window CGWindowID
  --expect-top-window <n>    pin baseline top-window CGWindowID
  --expect-space <n>         pin baseline active Space id (space-ctl active-space)
  --expect-focused-pid <n>   pin baseline focused-element owner pid
  --strict-focus             count AX focused-element changes as theft
  --min-samples <n>          post-settle samples required for ok (default 10)
  --report <path>            write the JSON report (default <cwd>/focus-probe-report.json)
  --samples-out <path>       also write the raw NDJSON sample stream
  --label <text>             label recorded in the report (default "cli")`;

const fail = (message: string): never => {
  console.error(`focus-probe: ${message}\n\n${usage}`);
  process.exit(1);
};

function parseArgs(argv: string[]): CliOptions {
  const options: CliOptions = {
    binary: process.env.SYNARA_CUA_FOCUS_PROBE ?? defaultBinary,
    durationSeconds: null,
    wrap: null,
    hz: 50,
    settleMs: 0,
    expect: {},
    strictFocus: false,
    minSamples: 10,
    reportPath: null,
    samplesPath: null,
    label: "cli",
  };
  const integer = (name: string, value: string): number => {
    const parsed = Number(value);
    if (!Number.isSafeInteger(parsed)) fail(`${name} needs an integer, got ${value}`);
    return parsed;
  };
  for (let index = 0; index < argv.length; index++) {
    const argument = argv[index]!;
    const value = () => {
      if (index + 1 >= argv.length) fail(`${argument} needs a value`);
      return argv[++index]!;
    };
    switch (argument) {
      case "--binary":
        options.binary = value();
        break;
      case "--duration":
        options.durationSeconds = Number(value());
        if (!(options.durationSeconds > 0)) fail("--duration needs a positive number");
        break;
      case "--wrap":
        if (argv[index + 1] !== "--") fail("--wrap must be followed by -- and a command");
        index += 1;
        options.wrap = argv.slice(index + 1);
        if (options.wrap.length === 0) fail("--wrap needs a command after --");
        return options;
      case "--hz":
        options.hz = Number(value());
        if (!(options.hz >= 1 && options.hz <= 500)) fail("--hz out of range (1-500)");
        break;
      case "--settle-ms":
        options.settleMs = integer("--settle-ms", value());
        break;
      case "--expect-frontmost-pid":
        options.expect.pid = integer("--expect-frontmost-pid", value());
        break;
      case "--expect-key-window":
        options.expect.keyWin = integer("--expect-key-window", value());
        break;
      case "--expect-top-window":
        options.expect.topWin = integer("--expect-top-window", value());
        break;
      case "--expect-space":
        options.expect.space = integer("--expect-space", value());
        break;
      case "--expect-focused-pid":
        options.expect.focusedPid = integer("--expect-focused-pid", value());
        break;
      case "--strict-focus":
        options.strictFocus = true;
        break;
      case "--min-samples":
        options.minSamples = integer("--min-samples", value());
        break;
      case "--report":
        options.reportPath = value();
        break;
      case "--samples-out":
        options.samplesPath = value();
        break;
      case "--label":
        options.label = value();
        break;
      case "--help":
        console.log(usage);
        process.exit(0);
      default:
        fail(`unknown argument: ${argument}`);
    }
  }
  if (!options.durationSeconds && !options.wrap) fail("pass --duration or --wrap");
  return options;
}

function ensureBinary(path: string): void {
  if (existsSync(path)) return;
  if (path !== defaultBinary)
    fail(
      `probe binary missing at ${path} (build with scripts/computer-use-fixtures/build-focus-probe.sh)`,
    );
  const source = join(root, "scripts/computer-use-fixtures/focus_probe.m");
  const built = spawnSync(
    "/usr/bin/clang",
    [
      "-fobjc-arc",
      "-O2",
      "-o",
      path,
      source,
      "-framework",
      "AppKit",
      "-framework",
      "ApplicationServices",
      "-framework",
      "CoreGraphics",
    ],
    { stdio: ["ignore", "inherit", "inherit"] },
  );
  if (built.status !== 0 || !existsSync(path))
    fail(`probe binary missing and clang build failed (status ${built.status ?? "spawn error"})`);
  console.error(`focus-probe: built ${path}`);
}

async function main(): Promise<number> {
  const options = parseArgs(process.argv.slice(2));
  ensureBinary(options.binary);

  const args = ["--hz", String(options.hz), "--label", options.label];
  // --duration runs are self-terminated by the sampler; --wrap holds stdin
  // open and closes it when the wrapped command exits.
  if (options.wrap) args.push("--stdin");
  else args.push("--duration", String(options.durationSeconds));
  const sampler: ChildProcessWithoutNullStreams = spawn(options.binary, args, {
    stdio: ["pipe", "pipe", "pipe"],
  });

  const samples: FocusProbeSample[] = [];
  const rawLines: string[] = [];
  let meta: FocusProbeMeta | null = null;
  let done: FocusProbeDone | null = null;
  let stderr = "";
  let buffer = "";
  sampler.stdout.on("data", (chunk: Buffer) => {
    buffer += chunk.toString("utf8");
    for (;;) {
      const end = buffer.indexOf("\n");
      if (end < 0) break;
      const line = buffer.slice(0, end);
      buffer = buffer.slice(end + 1);
      if (options.samplesPath) rawLines.push(line);
      const parsed = parseFocusProbeLine(line);
      if (!parsed) continue;
      if (parsed.kind === "meta") meta = parsed.meta;
      else if (parsed.kind === "done") done = parsed.done;
      else samples.push(parsed.sample);
    }
  });
  sampler.stderr.on("data", (chunk: Buffer) => {
    if (stderr.length < 16_384) stderr += chunk.toString("utf8");
  });
  const samplerExited = new Promise<number | null>((resolve) => {
    sampler.once("error", () => resolve(null));
    sampler.once("exit", (code) => resolve(code));
  });

  let interrupted = false;
  let wrapped: ChildProcessWithoutNullStreams | null = null;
  const onSignal = () => {
    interrupted = true;
    wrapped?.kill("SIGTERM");
    sampler.stdin.destroy();
    sampler.kill("SIGTERM");
  };
  process.once("SIGINT", onSignal);
  process.once("SIGTERM", onSignal);

  let wrappedExit: number | null = null;
  if (options.wrap) {
    wrapped = spawn(options.wrap[0]!, options.wrap.slice(1), {
      stdio: "inherit",
      env: process.env,
    });
    wrappedExit = await new Promise<number | null>((resolve) => {
      wrapped!.once("error", () => resolve(null));
      wrapped!.once("exit", (code) => resolve(code));
    });
    sampler.stdin.end();
  }

  const killTimer = setTimeout(() => sampler.kill("SIGKILL"), 15_000);
  const samplerExit = await samplerExited;
  clearTimeout(killTimer);

  const report = analyzeFocusSamples(samples, {
    settleMs: options.settleMs,
    expect: options.expect,
    strictFocus: options.strictFocus,
    minSamples: options.minSamples,
  });

  const reportPath = options.reportPath ?? join(process.cwd(), "focus-probe-report.json");
  await mkdir(dirname(reportPath), { recursive: true });
  await writeFile(
    reportPath,
    JSON.stringify(
      {
        probe: "focus-probe",
        label: options.label,
        options: {
          binary: options.binary,
          hz: options.hz,
          settleMs: options.settleMs,
          expect: options.expect,
          strictFocus: options.strictFocus,
          minSamples: options.minSamples,
          wrap: options.wrap,
          durationSeconds: options.durationSeconds,
        },
        meta,
        done,
        samplerExit,
        wrappedExit,
        interrupted,
        stderr: stderr.trim() || undefined,
        report,
      },
      null,
      2,
    ) + "\n",
  );
  if (options.samplesPath) {
    await mkdir(dirname(options.samplesPath), { recursive: true });
    await writeFile(options.samplesPath, rawLines.join("\n") + (rawLines.length ? "\n" : ""));
  }

  const summary = {
    theftFree: report.theftFree,
    ok: report.ok,
    samples: report.sampleCount,
    measuredMs: Math.round(report.measuredMs),
    baseline: report.baseline,
    offBaseline: report.offBaseline.length,
    offBaselineSamples: report.offBaselineSamples,
    drift: report.driftSamples,
    uncertainSamples: report.uncertainSamples,
    coverage: report.coverage,
    issues: report.issues,
    reportPath,
  };
  console.log(JSON.stringify(summary, null, 2));

  if (!report.theftFree) return 2;
  // An interrupted run is not a completed clean measurement even when the
  // partial samples pass the thresholds.
  if (interrupted) return 1;
  if (!report.ok) return 1;
  return 0;
}

main()
  .then((code) => process.exit(code))
  .catch((error: unknown) => {
    console.error(`focus-probe: ${String(error)}`);
    process.exit(1);
  });
